Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+  · rw [singleton_subset_iff]
+    rcases IsLUB.nonempty h_supX with ⟨x, xX⟩
+    have : infX = x := LE.le.antisymm (h_infX.1 xX) (h_supX.1 xX)
+    exact mem_of_eq_of_mem (id this) xX
+
+end inf_sup_section
+end conn_section
+end X_section
+
+/-- A tag for the tagged union of kinds of bounded intervals -/
+inductive BoundedIntervalKind | IooKind | IocKind | IcoKind | IccKind
+
+open BoundedIntervalKind
+
+/-- A tagged union of the kinds of nontrivial bounded intervals: `Ioo`, `Ioc`, `Ico`, and `Icc`. -/
+structure BoundedInterval :=
+  /-- the left endpoint (supremum) of the interval -/
+  left_endpoint : ℝ
+  /-- the right endpoint (infimum) of the interval -/
+  right_endpoint : ℝ
+  /-- a proof that the interval is nontrivial -/
+  left_lt_right : left_endpoint < right_endpoint
+  /-- tag indicating which endpoints are included -/
+  kind : BoundedIntervalKind
+
+/-- Convert the data of an interval to the set of real numbers it represents -/
+def BoundedInterval_as_set : BoundedInterval → Set ℝ
+  | ⟨a, b, _, IooKind⟩ => Ioo a b
+  | ⟨a, b, _, IocKind⟩ => Ioc a b
+  | ⟨a, b, _, IcoKind⟩ => Ico a b
+  | ⟨a, b, _, IccKind⟩ => Icc a b
+
+/-- True iff X is one of Ioo a b, Ioc a b, Ico a b, or Icc a b for some a < b-/
+def isBoundedInterval (X : Set ℝ) :=
+  ∃ I : BoundedInterval, X = BoundedInterval_as_set I
+
+lemma isBoundedInterval_Ioo {a b : ℝ} (lt : a < b) : isBoundedInterval (Ioo a b) := by
+  use ⟨a, b, lt, IooKind⟩; rw [BoundedInterval_as_set]
+lemma isBoundedInterval_Ioc {a b : ℝ} (lt : a < b) : isBoundedInterval (Ioc a b) := by
+  use ⟨a, b, lt, IocKind⟩; rw [BoundedInterval_as_set]
+lemma isBoundedInterval_Ico {a b : ℝ} (lt : a < b) : isBoundedInterval (Ico a b) := by
+  use ⟨a, b, lt, IcoKind⟩; simp only [BoundedInterval_as_set]
+lemma isBoundedInterval_Icc {a b : ℝ} (lt : a < b) : isBoundedInterval (Icc a b) := by
+  use ⟨a, b, lt, IccKind⟩; simp only [BoundedInterval_as_set]
+
+/-- The major lemma, assuming inf and sup exist. -/
+lemma classify_connected_reals_with_GLB_lt_LUB
+    {X : Set ℝ} (conn : IsConnected X) {infX supX : ℝ}
+    (h_infX : IsGLB X infX) (h_supX : IsLUB X supX) (lt : infX < supX) :
+    isBoundedInterval X := by
+  by_cases infX_X : infX ∈ X
+  · by_cases supX_X : supX ∈ X
+    · use ⟨infX, supX, lt, IccKind⟩
+      exact characterize_Icc conn h_infX h_supX infX_X supX_X
+    · use ⟨infX, supX, lt, IcoKind⟩
+      exact characterize_Ico conn h_infX h_supX lt infX_X supX_X
+  · by_cases supX_X : supX ∈ X
+    · use ⟨infX, supX, lt, IocKind⟩
+      exact characterize_Ioc conn h_infX h_supX lt infX_X supX_X
+    · use ⟨infX, supX, lt, IooKind⟩
+      exact characterize_Ioo conn h_infX h_supX infX_X supX_X
+
+/-- A connected subset of ℝ is either a singleton or a nontrivial bounded interval. -/
+theorem classify_connected_bounded_reals
+    {X : Set ℝ} (conn : IsConnected X) (above : BddAbove X) (below : BddBelow X) :
+    (∃ a : ℝ, X = {a}) ∨ isBoundedInterval X := by
+  have nonempty := IsConnected.nonempty conn
```

Comment:
You can use dot notation again: `conn.nonempty`. (And you may consider inlining this `have` into the two lemmas below.)